File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height)-1
        maxWater = min(height[l], height[r]) * (r - l)
        while l < r:
            if height[l] < height[r]:
                newL = l + 1
                while height[newL] <= height[l] and newL < r:
                    newL += 1
                l = newL
            else:
                newR = r - 1
                while height[newR] <= height[r] and l < newR:
                    newR -= 1
                r = newR
            water = min(height[l], height[r]) * (r - l)
            if water > maxWater:
                maxWater = water
        return maxWater

File: 11-container-with-most-water/container-with-most-water.ts
function maxArea(height: number[]): number {
    const n = height.length;
    let left = 0;
    let right = n-1;
    let max = Math.min(height[left], height[right]) * (right-left);
    if (height[left] < height[right]) {
        left++;
    } else {
        right--;
    }
    while (left < right) {
        max = Math.max(max, Math.min(height[left], height[right]) * (right-left));
        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }
    return max;
};